Link most "Partition II" links present in Partition I, and in III.4

darthwalsh opened the pull request:
In `ldfld` the text `[Partition II]` had a broken markdown link, so fix that.

Did a text search in both markdown and PDF, so found a couple missing sections (#9), only if they had a "Partition II" text. Clicked the links to find where they goes (although maybe there's a smarter way?)

"Partition II" in i.10.7.2-type-names-and-arity-encoding.md isn't a link (might be a spec mistake); remove link.

Didn't change link in these files because it links to the very start "Partition II", which isn't an obvious markdown target. Could link into `TABLE_OF_CONTENTS.md` or `ii.1-introduction.md`?
- i.6.2.1-managed-code.md
- i.8.2.2-built-in-value-and-reference-types.md
- i.8.9.1-array-types.md
- i.9-metadata.md
- i.9.2-accessing-metadata.md
- i.10.2-overloading.md

Also fix newline in i.6.2.1-managed-code.md

darthwalsh committed:
Fix ldfld broken link syntax

In the ldfld the text `[Partition II]` had a broken markdown link.
PDF links to II.22, but link a little deeper to the field table.

darthwalsh commented:
(In the PR, the link doesn't work in the PR view of the file preview.)
Go to https://github.com/darthwalsh/ecma-335/blob/darthwalsh-fix-link/docs/iii.4.10-ldfld.md and you can see the link works.

darthwalsh commented:
@stakx sorry to make two PRs on the same day. I now rebased this branch onto `master`.

stakx commented:
@darthwalsh, no need to apologize for sending PRs in parallel. I am happy & excited to see contributions!

stakx commented on docs/iii.4.10-ldfld.md:
```diff
@@ -10,7 +10,7 @@
 #### Description:
 
-The `ldfld` instruction pushes onto the stack the value of a field of _obj_. _obj_ shall be an object (type `O`), a managed pointer (type `&`), an unmanaged pointer (type `native int`), or an instance of a value type. The use of an unmanaged pointer is not permitted in verifiable code. _field_ is a metadata token (a `fieldref` or `fielddef` see [Partition II]#todo-missing-hyperlink)) that shall r ...
+The `ldfld` instruction pushes onto the stack the value of a field of _obj_. _obj_ shall be an object (type `O`), a managed pointer (type `&`), an unmanaged pointer (type `native int`), or an instance of a value type. The use of an unmanaged pointer is not permitted in verifiable code. _field_ is a metadata token (a `fieldref` or `fielddef` see [Partition II](ii.22.15-field-0x04.md)) that shall r ...
```
> PDF links to II.22, but link a little deeper to the field table.

I appreciate that you took the time to investigate possible link targets. However, I'd like to request that you change this to a hyperlink to II.22 (`ii.22-metadata-logical-format-tables.md`), for two reasons:

1. We'd stay closer to the original material (like you mentioned, the PDF links to II.22). This is important because this project's primary goal is to translate the original document to the Markdown format; this project shouldn't try to "fix" the standard, even where it's inaccurate.

2. While II.22 doesn't explicitly define `fielddef` and `fieldref` (as mentioned e.g. here), it does define _FieldRef_ (as a token into the _MemberRef_ metadata table) as well as introduce the metadata tables (_Field_ among them), so in a way it covers both of those field references. Linking to II.22.15 OTOH would only cover _FieldDef_ but not _FieldRef_.

darthwalsh replied:
Looks like you dug deeper ;) I'll update these links.

> this project's primary goal is to translate the original document to the Markdown format

Thanks for the clarity, this makes total sense! 

On a related tangent, have you seen this page: https://github.com/dotnet/runtime/blob/main/docs/design/specs/Ecma-335-Augments.md

Microsoft seems to have [moved away from updating specs](https://github.com/microsoft/TypeScript/issues/15711#issuecomment-661290678) so it might be interesting to incorporate the diffs that the dotnet team has decided on (maybe as a branch rebased ahead of `master`) 

Anyways, that's enough spitballing from me for now.

stakx replied:
I appreciate your input!

> On a related tangent, have you seen this page: https://github.com/dotnet/runtime/blob/main/docs/design/specs/Ecma-335-Augments.md

Yes, I've seen those kinds of "augment" documents before. I should add this one to the list (see the open issues at this repository). Thanks for mentioning it.

> Microsoft seems to have [moved away from updating specs](https://github.com/microsoft/TypeScript/issues/15711#issuecomment-661290678) so it might be interesting to incorporate the diffs that the dotnet team has decided on (maybe as a branch rebased ahead of master)

That's a fine idea. Adding their content to this project in dedicated branches would allow us to switch to a different branch and see the augmented/changed information, possibly marked up using a ` ```diff -... +... ``` ` so the changes are easier to spot.

stakx commented on docs/iii.4.10-ldfld.md:
```diff
@@ -10,7 +10,7 @@
 #### Description:
 
-The `ldfld` instruction pushes onto the stack the value of a field of _obj_. _obj_ shall be an object (type `O`), a managed pointer (type `&`), an unmanaged pointer (type `native int`), or an instance of a value type. The use of an unmanaged pointer is not permitted in verifiable code. _field_ is a metadata token (a `fieldref` or `fielddef` see [Partition II]#todo-missing-hyperlink)) that shall r ...
+The `ldfld` instruction pushes onto the stack the value of a field of _obj_. _obj_ shall be an object (type `O`), a managed pointer (type `&`), an unmanaged pointer (type `native int`), or an instance of a value type. The use of an unmanaged pointer is not permitted in verifiable code. _field_ is a metadata token (a `fieldref` or `fielddef` see [Partition II](ii.22.15-field-0x04.md)) that shall r ...
```
Btw. there are a few other places with identical references, where the hyperlink is also still missing. While we're at it, could you fix those hyperlinks too?

* in [III.4.11](https://github.com/stakx/ecma-335/blob/master/docs/iii.4.11-ldflda.md)
* in [III.4.14](https://github.com/stakx/ecma-335/blob/master/docs/iii.4.14-ldsfld.md)
* in [III.4.15](https://github.com/stakx/ecma-335/blob/master/docs/iii.4.15-ldsflda.md)
* in [III.4.28](https://github.com/stakx/ecma-335/blob/master/docs/iii.4.28-stfld.md)
* in [III.4.30](https://github.com/stakx/ecma-335/blob/master/docs/iii.4.30-stsfld.md)

darthwalsh replied:
👍  made those changes, and went a little further :D

stakx reviewed (changes requested):
Thanks once again for your PR. I'd be happy if we could change those missing hyperlinks to links to II.22 (as per the PDF).

darthwalsh committed:
Update [Partition II] links in files matching /iii.4*/

Tested all links in the PDF and they all went to II.22

darthwalsh committed:
Link most "Partition II" links present in Partition I

Did a text search in both markdown and PDF, so found a couple #9 missing sections, only if they had a "Partition II" text. Clicked the links to find where they goes (although maybe there's a smarter way?)

"Partition II" in i.10.7.2-type-names-and-arity-encoding.md isn't a link (might be a spec mistake); remove link.

Didn't change link in these files because it links to the very start "Partition II", which isn't an obvious markdown target. Could link into `TABLE_OF_CONTENTS.md` or `ii.1-introduction.md`?
- i.6.2.1-managed-code.md
- i.8.2.2-built-in-value-and-reference-types.md
- i.8.9.1-array-types.md
- i.9-metadata.md
- i.9.2-accessing-metadata.md
- i.10.2-overloading.md

Also fix newline in i.6.2.1-managed-code.md

darthwalsh commented:
@stakx  not sure if this PR is still on your radar, but I think it's still ready to merge

stakx commented on docs/i.10.2-overloading.md:
```diff
@@ -38,4 +38,4 @@ Properties shall not be overloaded by type (that is, by the return type of their
 > **CLS (framework):** Shall not publicly expose overloading except as specified here. Frameworks authors should bear in mind that many programming languages, including object-oriented languages, do not support overloading and will expose overloaded methods or properties through mangled names. Most languages support neither operator overloading nor overloading based on return type, so `op_Implici ...
 
-_[Note:_ The names visible on any class `C`, are the names visible in that class and its base classes. As a consequence, the names of methods on interfaces implemented by `C` that are only implemented via MethodImpls (see [Partition II](#todo-missing-hyperlink)) are not visible on class `C`. The names visible on an interface `I`, consist only of the names directly defined on this interface. As a  ...
+_[Note:_ The names visible on any class `C`, are the names visible in that class and its base classes. As a consequence, the names of methods on interfaces implemented by `C` that are only implemented via MethodImpls (see [Partition II](ii.25-file-format-extensions-to-pe.md)) are not visible on class `C`. The names visible on an interface `I`, consist only of the names directly defined on this in ...
```
All good. Just noting that I verified that the original PDF indeed goes to II.25, when II.22.27 MethodImpl would be the more logical hyperlink target.

stakx commented on docs/i.11-collected-common-language-specification-rules.md:
```diff
@@ -70,7 +70,7 @@ The complete set of CLS rules are collected here for reference. Recall that thes
 **CLS Rule 34:** The CLS only allows a subset of the encodings of custom attributes.  The only types that shall appear in these encodings are (see Partition IV): `System.Type`, `System.String`, `System.Char`, `System.Boolean`, `System.Byte`, `System.Int16`, `System.Int32`, `System.Int64`, `System.Single`, `System.Double`, and any enumeration type based on a CLS-compliant base integer type. (§[I.9 ...
 
-**CLS Rule 35:** The CLS does not allow publicly visible required modifiers (modreq, see [Partition II](#todo-missing-hyperlink)), but does allow optional modifiers (modopt, see [Partition II](#todo-missing-hyperlink)) it does not understand. (§[I.9.7](i.9.7-metadata-extensibility.md))
+**CLS Rule 35:** The CLS does not allow publicly visible required modifiers (modreq, see [Partition II](i.9.7-metadata-extensibility.md#cls-rule-35)), but does allow optional modifiers (modopt, see [Partition II](i.9.7-metadata-extensibility.md#cls-rule-35)) it does not understand. (§[I.9.7](i.9.7-metadata-extensibility.md))
```
I realise that if one clicks on the "Partition II" text in the original PDF, one ends up at `i.9.7-metadata-extensibility.md`, where CLS Rule 35 gets introduced; and there you have the same hyperlinks, but this time they go to `ii.7.1.1-modreq-and-modopt.md`.

I'm undecided, but wouldn't it be better to have the hyperlinks here go directly to `ii.7.1.1-modreq-and-modopt.md`, since this chapter essentially contains a copy of all the CLS Rules introduced earlier?

stakx commented on docs/i.12.2-module-information.md:
```diff
@@ -1,12 +1,12 @@
  * The _signature_ of the method, which specifies the return type and the number, order, parameter passing convention, and built-in data type of each of the arguments. It also specifies the native calling convention (this does not affect the CIL virtual calling convention, just the native code).
 
- * The _exception handling array_. This array holds information delineating the ranges over which exceptions are filtered and caught. See [Partition II](#todo-missing-hyperlink) and §[I.12.4.2](i.12.4.2-exception-handling.md).
+ * The _exception handling array_. This array holds information delineating the ranges over which exceptions are filtered and caught. See [Partition II](i.12.4.2-exception-handling.md) and §[I.12.4.2](i.12.4.2-exception-handling.md).
```
Weird, but correct: the original PDF links this "Partition II" link with Partition I.

stakx commented on docs/i.6.2.1-managed-code.md:
```diff
@@ -8,4 +8,6 @@
- * Store and retrieve security information This standard specifies a particular instruction set, the CIL (see [Partition III](#todo-missing-hyperlink)), and a file format (see [Partition II](#todo-missing-hyperlink)) for storing and transmitting managed code.
+ * Store and retrieve security information
+ 
+ This standard specifies a particular instruction set, the CIL (see [Partition III](#todo-missing-hyperlink)), and a file format (see [Partition II](#todo-missing-hyperlink)) for storing and transmitting managed code.
```
I'm assuming you didn't change this "Partition II" hyperlink because the PDF just points at the beginning of the partition, and we cannot pinpoint that more precisely (II.24 and II.25 would _both_ be candidates but we could only link to one of those). I'm fine with leaving this hyperlink broken for now.

stakx commented on docs/i.9.2-accessing-metadata.md:
```diff
@@ -1,5 +1,5 @@
 ## I.9.2 Accessing metadata
 
-Metadata is emitted into and read from a CLI module using either direct access to the file format as described in [Partition II Metadata &ndash; File Format](#todo-missing-hyperlink) or through the Reflection library. It is possible to create a tool that verifies a CLI module, including the metadata, during development, based on the specifications supplied in [Partition II](#todo-missing-hyperlin ...
+Metadata is emitted into and read from a CLI module using either direct access to the file format as described in [Partition II Metadata &ndash; File Format](ii.25-file-format-extensions-to-pe.md) or through the Reflection library. It is possible to create a tool that verifies a CLI module, including the metadata, during development, based on the specifications supplied in [Partition II](#todo-mi ...
```
Assuming (again) that you left this "Partition II" hyperlink broken because the PDF just links to the start of Partition II, which isn't very specific. I'm OK with leaving it broken for now.

stakx reviewed (approved):
Sorry for the very long radio silence. I haven't been very active on GitHub lately due to being rather busy at my daytime job. I needed a break from programming-related activities in my spare time.

That being said: You did a near perfect job here :100:, thanks so much for taking the time to go through the whole thing and fixing all those "Partition II" links.

stakx commented:
> Didn't change link in these files because it links to the very start "Partition II", which isn't an obvious markdown target. Could link into `TABLE_OF_CONTENTS.md` or `ii.1-introduction.md`?
>
> * i.6.2.1-managed-code.md
> * i.8.2.2-built-in-value-and-reference-types.md
> * i.8.9.1-array-types.md
> * i.9-metadata.md
> * i.9.2-accessing-metadata.md
> * i.10.2-overloading.md

Since I let so much time pass, I didn't want to hold back your PR any further with additional requests. But if you ever find yourself wanting to contribute to this repository again, I think linking to the beginning of Partition II in `TABLE_OF_CONTENTS.md` would perhaps be the most helpful for readers.
